common/tablefunctions.lua

```diff
@@ -310,14 +310,15 @@ end
 
 if not table.map then
 	--- Applies a function to all elements of a table and returns a new table with the results.
-	---@generic K, V, R
+	---@generic K, V, RV, RK
 	---@param tbl table<K, V> The input table.
-	---@param callback fun(value: V, key: K, tbl: table<K, V>): R The function to apply to each element. It receives three arguments: the element's value, its key, and the original table.
-	---@return R[] A new table containing the results of applying the callback to each element.
+	---@param callback fun(value: V, key: K, tbl: table<K, V>): RV, RK The function to apply to each element. It receives three arguments: the element's value, its key, and the original table. It should return the new value, and optionally, a new key.
+	---@return table<RK, RV> A new table containing the results of applying the callback to each element.
 	function table.map(tbl, callback)
 		local result = {}
 		for k, v in pairs(tbl) do
-			result[k] = callback(v, k, tbl)
+			local mv, mk = callback(v, k, tbl)
+			result[mk ~= nil and mk or k] = mv
```

Comment:
why not just `mk or k`? if `mk` is `false` it will use `k` anyway

Reply:
Ah, you're right. I think I'll have to change it to a proper `if`. I'm used to using `...and...or...` as a ternary operator, but it's not quite the same.

Reply:
fixed